torch/distributed/_shard/checkpoint/storage.py

```diff
@@ -0,0 +1,167 @@
+import abc
+from typing import List, Dict
+
+from torch.futures import Future
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+
+class StorageWriter(abc.ABC):
+    """
+    Interface used by ``save_state_dict`` to write to storage.
+
+    A subclass should expect the following sequence of calls by ``save_state_dict``
+
+    1) prepare_storage with a dict of storage_key -> size. This can be used to
+        pre-allocate storage.
+    2) write_bytes
+    3) write_tensors.
+    4) (if rank == 0) write_metadata().
+
+    Rank 0 will wait for (2) and (3) before writing metadata.
+    The writes from (2) and (3) are initiated before any waiting is done.
+    """
+
+    @abc.abstractmethod
+    def write_bytes(self, requests: List[BytesWriteRequest]) -> Future[None]:
+        """
+        Initiate writes for all requests in `requests`.
+
+        Writing can happen asynchronously and/or concurrently. A blocking
+        implementation is valid.
+
+        Args:
+            requests (List[BytesWriteRequest]): A list of requests to write
+        Returns:
+            A future that completes once all writes have finished.
+        """
+        pass
+
+    @abc.abstractmethod
+    def write_tensors(self, requests: List[TensorWriteRequest]) -> Future[None]:
+        """
+        Initiate writes for all requests in `requests`.
+
+        Writing can happen asynchronously and/or concurrently. A blocking
+        implementation is valid.
+
+        Implementors are responsible for any device to host transfers required
+        to copy.
+
+        Args:
+            requests (List[TensorWriteRequest]): A list of requests to write
+
+        Returns:
+            A future that completes once all writes have finished.
+        """
+        pass
+
+    @abc.abstractmethod
+    def write_metadata(self, metadata: Metadata) -> None:
```

Comment:
I'm updating the docs to be clear that `write_metadata` does commit the checkpoint.

On the second part of your comment, is this about a StorageWriter impl having freedom to use any schema/format for the data or whether any data needs to be written at all.

An example of my later point is a storage layer where each rank saves the metadata along the data they are writing and the loader consumes that to produce the global metadata object.

Is this sort of scenario that you had in mind?

While I agree with you that this needs to be addressed, this can be done in a follow up change. I tracking this in https://github.com/pytorch/pytorch/issues/76441